File: solutions/2026-07-01/cpp/2026-07-01-find-the-safest-path-in-a-grid.cpp
```cpp
// Problem: Find the Safest Path in a Grid
// Link: https://leetcode.com/problems/find-the-safest-path-in-a-grid/
//
// Approach:
// The problem asks for the maximum safeness factor of a path from (0,0) to (n-1, n-1).
// The safeness factor of a path is the minimum Manhattan distance from any cell in the path to any thief.
// This problem can be rephrased as finding a path from (0,0) to (n-1, n-1) such that the minimum distance
// from any cell on the path to *any* thief is maximized.
//
// This suggests a binary search approach on the answer (the maximum safeness factor).
// For a given potential safeness factor 'k', we need to check if there exists a path from (0,0) to (n-1, n-1)
// where every cell in the path is at a Manhattan distance of at least 'k' from any thief.
//
// To efficiently check this, we first precompute the minimum Manhattan distance from every cell to the nearest thief.
// This can be done using a multi-source BFS. Initialize a distance matrix with infinity for all cells,
// and 0 for cells containing thieves. Add all thief cells to the BFS queue.
//
// After precomputing the distances, for a given 'k', we can perform a BFS or DFS on the grid.
// We are only allowed to visit cells (r, c) if their precomputed minimum distance to a thief is >= 'k'.
// If we can reach (n-1, n-1) from (0,0) using only such cells, then a path with safeness factor 'k' exists.
//
// The binary search will work as follows:
// Lower bound for safeness factor is 0.
// Upper bound can be estimated as n-1 (maximum possible Manhattan distance in an n x n grid).
// For a 'mid' value in binary search:
//   - If a path exists with safeness 'mid', it means we can achieve at least 'mid', so we try for a higher safeness factor (low = mid + 1).
//   - If no path exists with safeness 'mid', we must reduce our target safeness factor (high = mid - 1).
//
// Precomputation of distances:
// We can use a BFS starting from all thief cells simultaneously.
// Let `dist[r][c]` be the minimum Manhattan distance from cell (r, c) to any thief.
// Initialize `dist` with -1 (or infinity) for all cells.
// For each cell (r, c) with `grid[r][c] == 1`, set `dist[r][c] = 0` and add it to the BFS queue.
// When processing a cell (r, c) from the queue, for each of its neighbors (nr, nc):
//   If `dist[nr][nc]` is still -1, update `dist[nr][nc] = dist[r][c] + 1` and add (nr, nc) to the queue.
//
// Path existence check (for a given 'k'):
// Use BFS. Start BFS from (0,0).
// Queue stores cells to visit.
// `visited` array to keep track of visited cells during this BFS.
// Add (0,0) to queue if `dist[0][0] >= k`. Mark (0,0) as visited.
// While queue is not empty:
//   Dequeue (r, c).
//   If (r, c) is (n-1, n-1), return true.
//   For each neighbor (nr, nc):
//     If (nr, nc) is within bounds, not visited, and `dist[nr][nc] >= k`:
//       Enqueue (nr, nc). Mark as visited.
// If queue becomes empty and (n-1, n-1) not reached, return false.
//
// Time Complexity:
// 1. Precomputing distances: Multi-source BFS takes O(n*n) time, where n is the grid dimension.
// 2. Binary search: The range of safeness factor is from 0 to O(n). Let's say log(max_dist) iterations.
// 3. Path existence check (BFS): In each iteration of binary search, a BFS takes O(n*n) time.
// Total time complexity: O(n*n * log(n)).
//
// Space Complexity:
// 1. Distance matrix: O(n*n)
// 2. Visited matrix for path check: O(n*n)
// 3. Queue for BFS: O(n*n) in the worst case.
// Total space complexity: O(n*n).

#include <vector>
#include <queue>
#include <algorithm>

using namespace std;

class Solution {
public:
    // Directions for moving to adjacent cells (up, down, left, right)
    int dr[4] = {-1, 1, 0, 0};
    int dc[4] = {0, 0, -1, 1};

    // Function to check if a given safeness factor 'k' is achievable.
    // This function performs a BFS to see if there's a path from (0,0) to (n-1, n-1)
    // where all cells on the path have a minimum distance to a thief of at least 'k'.
    bool canReach(int k, int n, const vector<vector<int>>& dist) {
        // If the starting cell (0,0) itself is not safe enough, we can't even start.
        if (dist[0][0] < k) {
            return false;
        }

        // BFS queue to store cells to visit. Each element is a pair {row, col}.
        queue<pair<int, int>> q;
        // Visited matrix to keep track of visited cells during this BFS to avoid cycles.
        vector<vector<bool>> visited(n, vector<bool>(n, false));

        // Start BFS from (0,0)
        q.push({0, 0});
        visited[0][0] = true;

        while (!q.empty()) {
            pair<int, int> current_cell = q.front();
            q.pop();
            int r = current_cell.first;
            int c = current_cell.second;

            // If we reached the destination cell (n-1, n-1), then 'k' is achievable.
            if (r == n - 1 && c == n - 1) {
                return true;
            }

            // Explore adjacent cells
            for (int i = 0; i < 4; ++i) {
                int nr = r + dr[i];
                int nc = c + dc[i];

                // Check if the neighbor is within grid bounds
                if (nr >= 0 && nr < n && nc >= 0 && nc < n) {
                    // Check if the neighbor has not been visited yet AND
                    // its minimum distance to a thief is at least 'k'
                    if (!visited[nr][nc] && dist[nr][nc] >= k) {
                        visited[nr][nc] = true; // Mark as visited
                        q.push({nr, nc});       // Add to the queue for exploration
                    }
                }
            }
        }

        // If the queue becomes empty and we haven't reached the destination,
        // then 'k' is not achievable with the current path constraints.
        return false;
    }

    // Main function to find the maximum safeness factor.
    int maximumSafenessPath(vector<vector<int>>& grid) {
        int n = grid.size();

        // dist[r][c] will store the minimum Manhattan distance from cell (r, c) to any thief.
        // Initialize with -1, signifying an unvisited or non-thief cell initially.
        vector<vector<int>> dist(n, vector<int>(n, -1));
        // BFS queue for precomputing distances. Stores {row, col}.
        queue<pair<int, int>> q;

        // Step 1: Precompute minimum Manhattan distances to the nearest thief for all cells.
        // Add all thief cells to the queue and set their initial distance to 0.
        for (int r = 0; r < n; ++r) {
            for (int c = 0; c < n; ++c) {
                if (grid[r][c] == 1) {
                    dist[r][c] = 0;
                    q.push({r, c});
                }
            }
        }

        // Multi-source BFS to calculate distances from all thieves simultaneously.
        while (!q.empty()) {
            pair<int, int> current_cell = q.front();
            q.pop();
            int r = current_cell.first;
            int c = current_cell.second;

            // Explore neighbors
            for (int i = 0; i < 4; ++i) {
                int nr = r + dr[i];
                int nc = c + dc[i];

                // Check if neighbor is within bounds and has not been visited yet (dist is -1).
                if (nr >= 0 && nr < n && nc >= 0 && nc < n && dist[nr][nc] == -1) {
                    // The distance to the neighbor is one more than the current cell's distance.
                    dist[nr][nc] = dist[r][c] + 1;
                    q.push({nr, nc}); // Add the neighbor to the queue for further exploration.
                }
            }
        }

        // Step 2: Binary search for the maximum safeness factor.
        // The safeness factor can range from 0 (if the path must pass through a thief cell)
        // up to n-1 (maximum possible Manhattan distance in an n x n grid).
        int low = 0;
        int high = n - 1; // A loose upper bound, could also be 2*(n-1) or even more. n-1 is sufficient.
        int max_safeness = 0;

        // Perform binary search
        while (low <= high) {
            int mid = low + (high - low) / 2; // Calculate mid to avoid overflow

            // Check if it's possible to reach (n-1, n-1) with a safeness factor of 'mid'.
            if (canReach(mid, n, dist)) {
                // If we can reach with 'mid', it means 'mid' is a possible safeness factor.
                // We try to achieve an even higher safeness factor.
                max_safeness = mid; // Store this as a potential answer
                low = mid + 1;      // Try for higher safeness
            } else {
                // If we cannot reach with 'mid', it means 'mid' is too high.
                // We need to reduce the target safeness factor.
                high = mid - 1;     // Try for lower safeness
            }
        }

        // The `max_safeness` variable will hold the highest value of 'mid' for which `canReach` returned true.
        return max_safeness;
    }
};
```

File: solutions/2026-07-01/python/2026-07-01-find-the-safest-path-in-a-grid.py
```python
# Finds the maximum safeness factor of a path from (0,0) to (n-1,n-1) in a grid.
# The safeness factor is the minimum Manhattan distance from any path cell to any thief.
# Link: https://leetcode.com/problems/find-the-safest-path-in-a-grid/
#
# Approach:
# 1. Precompute the minimum distance from each cell to the nearest thief using BFS.
#    We initialize a distance matrix with infinity and add all thief cells to the BFS queue
#    with distance 0. Then, we perform a multi-source BFS to populate the distance matrix.
# 2. The problem now transforms into finding a path from (0,0) to (n-1,n-1) such that
#    the minimum of these precomputed distances along the path is maximized.
# 3. This is a classic problem that can be solved using Binary Search on the answer (safeness factor).
#    For a given potential safeness factor 'k', we check if a path exists from (0,0) to (n-1,n-1)
#    where every cell on the path has a precomputed distance to a thief of at least 'k'.
# 4. The check for a given 'k' can be done using BFS or DFS. We only consider cells (r, c)
#    where `dist_to_thief[r][c] >= k`.
#
# Time Complexity:
# - Step 1 (Precomputing distances): O(N*N), where N is the dimension of the grid.
#   This is because each cell is visited at most once by the BFS.
# - Step 3 & 4 (Binary Search and Path Checking): The binary search performs log(MaxDistance) iterations.
#   The maximum possible Manhattan distance is 2*(N-1). So, log(N) iterations.
#   Inside each iteration, the path checking (BFS/DFS) takes O(N*N) time.
#   Therefore, the overall time complexity is O(N*N * log(N)).
#
# Space Complexity:
# - O(N*N) for storing the distance matrix and the BFS queue.

import collections
import heapq

class Solution:
    def maximumSafenessFactor(self, grid: list[list[int]]) -> int:
        n = len(grid)
        
        # Step 1: Precompute minimum distance from each cell to the nearest thief using BFS.
        # Initialize distances with infinity.
        dist_to_thief = [[float('inf')] * n for _ in range(n)]
        
        # Queue for BFS, storing (distance, row, col)
        q = collections.deque()
        
        # Add all thief locations to the queue with distance 0.
        for r in range(n):
            for c in range(n):
                if grid[r][c] == 1:
                    dist_to_thief[r][c] = 0
                    q.append((r, c))
        
        # Directions for BFS (up, down, left, right)
        dirs = [(0, 1), (0, -1), (1, 0), (-1, 0)]
        
        # Perform BFS to calculate minimum distances.
        while q:
            r, c = q.popleft()
            
            for dr, dc in dirs:
                nr, nc = r + dr, c + dc
                
                # Check if the neighbor is within grid bounds.
                if 0 <= nr < n and 0 <= nc < n:
                    # If the current path to the neighbor is shorter than its recorded distance, update it.
                    # The distance to the neighbor is one more than the current cell's distance.
                    if dist_to_thief[r][c] + 1 < dist_to_thief[nr][nc]:
                        dist_to_thief[nr][nc] = dist_to_thief[r][c] + 1
                        q.append((nr, nc))

        # Step 2: Binary Search for the maximum safeness factor.
        # The possible safeness factors range from 0 to 2*(n-1) (max Manhattan distance).
        low = 0
        high = 2 * (n - 1)
        max_safeness = 0
        
        # Iterate while the search space is valid.
        while low <= high:
            mid = (low + high) // 2 # Current potential safeness factor.
            
            # Step 3 & 4: Check if a path exists with a minimum safeness factor of 'mid'.
            # We use BFS for path checking.
            if self.can_reach_end(grid, dist_to_thief, mid):
                # If a path exists with safeness 'mid', it means we might be able to achieve a higher safeness.
                # Store 'mid' as a potential answer and try for higher values.
                max_safeness = mid
                low = mid + 1
            else:
                # If no path exists with safeness 'mid', we need to aim for a lower safeness factor.
                high = mid - 1
                
        return max_safeness

    def can_reach_end(self, grid: list[list[int]], dist_to_thief: list[list[int]], min_safeness: int) -> bool:
        """
        Checks if there is a path from (0,0) to (n-1,n-1) where every cell's
        distance to the nearest thief is at least 'min_safeness'.
        """
        n = len(grid)
        
        # If the starting or ending cell itself is unsafe (less than min_safeness),
        # then no path can exist.
        if dist_to_thief[0][0] < min_safeness or dist_to_thief[n-1][n-1] < min_safeness:
            return False
        
        # Queue for BFS, storing (row, col).
        q = collections.deque()
        # Set to keep track of visited cells to avoid cycles.
        visited = set()
        
        # Add the starting cell to the queue and mark as visited.
        q.append((0, 0))
        visited.add((0, 0))
        
        # Directions for BFS.
        dirs = [(0, 1), (0, -1), (1, 0), (-1, 0)]
        
        # Perform BFS.
        while q:
            r, c = q.popleft()
            
            # If we reached the destination, return True.
            if r == n - 1 and c == n - 1:
                return True
            
            for dr, dc in dirs:
                nr, nc = r + dr, c + dc
                
                # Check if the neighbor is within bounds, has not been visited,
                # and its distance to the nearest thief is at least 'min_safeness'.
                if (0 <= nr < n and 0 <= nc < n and
                    (nr, nc) not in visited and
                    dist_to_thief[nr][nc] >= min_safeness):
                    
                    # Add the valid neighbor to the queue and mark as visited.
                    q.append((nr, nc))
                    visited.add((nr, nc))
                    
        # If the BFS completes without reaching the destination, it means no such path exists.
        return False

```